azext_edge/edge/providers/rpsaas/base_provider.py

```diff
@@ -0,0 +1,262 @@
+            self.cmd,
+            subscription_id=custom_location_subscription,
+            custom_query=query,
+            type=ResourceTypeMapping.custom_location.value,
+            name=custom_location_name,
+            resource_group=custom_location_resource_group
+        )
+        if len(location_query_result) == 0:
+            error_details = ""
+            if custom_location_name:
+                error_details += f"{custom_location_name} "
+            if cluster_name:
+                error_details += f"for cluster {cluster_name} "
+            raise ResourceNotFoundError(f"Custom location {error_details}not found.")
+
+        if len(location_query_result) > 1 and cluster_name is None:
+            raise ValidationError(
+                MULTIPLE_POSSIBLE_ITEMS_ERROR.format(
+                    len(location_query_result),
+                    "custom location",
+                    custom_location_name
+                )
+            )
+        # by this point there should be at least one custom location
+        # if cluster name was given (and no custom_location_names), there can be more than one
+        # otherwise, if no cluster name, needs to be only one
+
+        # should trigger if only the location name was provided - there should be one location
+        if not cluster_name:
+            query = f'| where id =~ "{location_query_result[0]["properties"]["hostResourceId"]}"'
+            cluster_query_result = build_query(
+                self.cmd,
+                subscription_id=cluster_subscription,
+                custom_query=query,
+                type=ResourceTypeMapping.connected_cluster.value
+            )
+            if len(cluster_query_result) == 0:
+                raise ValidationError(CUSTOM_LOCATION_DOES_NOT_EXIST_ERROR.format(custom_location_name))
+            cluster = cluster_query_result[0]
+        # by this point, cluster is populated
+
+        if cluster["properties"]["connectivityStatus"].lower() != "connected":
+            logger.warning(CLUSTER_OFFLINE_MSG.format(cluster["name"]))
+
+        possible_locations = []
+        for location in location_query_result:
+            usable = False
+            for extension_id in location["properties"]["clusterExtensionIds"]:
+                # extensions not findable in graph :(
+                extension = self.resource_client.resources.get_by_id(
+                    resource_id=extension_id,
+                    api_version=EXTENSION_API_VERSION,
+                ).as_dict()
+                if extension["properties"]["extensionType"] == self.required_extension:
+                    usable = True
+                    break
+            if usable:
+                possible_locations.append(location["id"])
+
+        # throw if there are no suitable extensions (in the cluster)
+        if len(possible_locations) == 0:
+            raise ValidationError(MISSING_EXTENSION_ERROR.format(cluster["name"], self.required_extension))
```

Comment:
discussed offline about a potential tweak to making required_extension not optional